### ISSUE TITLE
vboxwrapper should use ncpus = physical cores

### ISSUE THREAD

**Issue by marius311** [BOINC/boinc#2542]
From https://www.virtualbox.org/manual/ch03.html:

> You should not, however, configure virtual machines to use more CPU cores than you have available physically (real cores, no hyperthreads).

However currently vboxwrapper uses the number logical cores. There is some discussion of this in https://github.com/BOINC/boinc/issues/1501 and https://github.com/marius311/boinc-server-docker/issues/33.

I had thought this was some minor sub-optimality, but thanks to @zonca getting a totally reproducible scenario working, I recently made the connection that this could be causing a significant number of the failures we see at C@H which I had previously not understood. 

I don't have a solution, see https://github.com/BOINC/boinc/issues/1501 for discussion of the issues involved in getting the number of physical cores (anyone know a good API?). I'm opening this though so we can brainstorm / track progress on this specific issue.

**Comment by AenBleidd** [comment]
I think this should be an option rather than just hard limitation.
In any case it's up to the user to choose how many VMs he want to be run.

**Comment by marius311** [comment]
I don't have any problem with it being overridable since indeed it seems in some cases it works fine with ncpus > physical cores, although the default should definitely be = physical cores.

**Comment by AenBleidd** [comment]
Yes, I agree with default value. I just want to be sure that this will be not hardcoded as max VM value = number of physical cores.

**Comment by lfield** [comment]
This seems similar to #2297. Are any code changes required?

**Comment by marius311** [comment]
ts possible that is the same issue, thanks for the link. 

The main code change is adding a way to detect the number of physical cores to the client and/or vboxwrapper itself (we don't currently know how to do this, although some leads were mentioned in the linked thread)

**Comment by marius311** [comment]
After some more digging, I think Linux and OSX may be fairly easy:

* Linux: `lscpu` gives you the total "CPU(s)" and the "Thread(s) per core", the physical cores is simply the first divided by the second. This has been correct on about 5 linux systems that I have tested that I have access to. Can / should we add `lscpu` as a dependency in this case? I think vast majority of Linux system come with it installed. We can probably parse the same info out of /proc/cpuinfo if not. 

* OSX: [This](https://coolaj86.com/articles/get-a-count-of-cpu-cores-on-linux-and-os-x/) would suggest its as easy as `sysctl -n hw.physicalcpu`. Can someone with a Mac test this for me? 

Windows may be more of a mess. See e.g. [here](https://stackoverflow.com/questions/2901694/programmatically-detect-number-of-physical-processors-cores-or-if-hyper-threadin). Does anyone with more Windows experience want to tackle this? It doesn't seem like its going to be totally pretty, but probably doable.

**Comment by AenBleidd** [comment]
I guess https://msdn.microsoft.com/en-us/library/ms683194 is definitely a solution for Windows.

**Comment by marius311** [comment]
Ah nice, I had missed that, thanks. Ok, great, in that case it definitely seems like we can progress. I'm happy to take the lead unless someone else wants to, but will definitely need some help, especially testing on Mac. 

For @AenBleidd, @davidpanderson, or anyone else that could offer advice. As a summary, the goal here is have the client log the number of physical cores, somehow I suppose this will make it back to the server as a new row in the host table, and then somehow we need to be able to specify jobs (I suppose via plan_classes) for which the max number of CPUs is the physical rather than the total number of CPUs. I admit I have almost no knowledge of how the scheduler, RPCs, etc.. work, if any of you could offer a high level overview of what the changes necessary to make this happen should look like, that'd be very helpful.

**Comment by CharlieFenton** [comment]
> OSX: This would suggest its as easy as sysctl -n hw.physicalcpu. Can someone with a Mac test this for me?

From _MacOSX10.13.sdk/usr/include/sys/sysctl.h_ and also found on the web [here](https://opensource.apple.com/source/xnu/xnu-792.12.6/bsd/sys/sysctl.h)
```
 * These are the support HW selectors for sysctlbyname.  Parameters that are byte counts or frequencies are 64 bit numbers.
 * All other parameters are 32 bit numbers.
 *
 *   hw.memsize                - The number of bytes of physical memory in the system.
 *
 *   hw.ncpu                   - The maximum number of processors that could be available this boot.
 *                               Use this value for sizing of static per processor arrays; i.e. processor load statistics.
 *
 *   hw.activecpu              - The number of processors currently available for executing threads.
 *                               Use this number to determine the number threads to create in SMP aware applications.
 *                               This number can change when power management modes are changed.
 *
 *   hw.physicalcpu            - The number of physical processors available in the current power management mode.
 *   hw.physicalcpu_max        - The maximum number of physical processors that could be available this boot
 *
 *   hw.logicalcpu             - The number of logical processors available in the current power management mode.
 *   hw.logicalcpu_max         - The maximum number of logical processors that could be available this boot

```
On my Retina 4K, 21.5-inch, 2017 iMac, `sysctl hw` gives me these values (among many others):
```
hw.ncpu: 4
hw.activecpu: 4
hw.physicalcpu: 4
hw.physicalcpu_max: 4
hw.logicalcpu: 4
hw.logicalcpu_max: 4

```
BOINC reports the processor as:
`Processor: 4 GenuineIntel Intel(R) Core(TM) i5-7400 CPU @ 3.00GHz [x86 Family 6 Model 158 Stepping 9]`
Since this processor does not support hyper threading, the number of logical cores equals the number of physical cores, as reported by _sysctl_.

**Comment by CharlieFenton** [comment]
Note that hostinfo_unix.cpp uses the C language APIs
` int sysctlbyname(const char *name, void *oldp, size_t *oldlenp, void *newp, size_t newlen);`
as described [here](http://www.manpagez.com/man/3/sysctlbyname/) rather than the command line interfaces like `sysctl -n hw.physicalcpu`.

Also worth noting is that the _man_ page for _sysctlbyname()_ says:
```
     hw.ncpu
             The number of cpus. This attribute is deprecated and it is recom-
             mended that hw.logicalcpu, hw.logicalcpu_max, hw.physicalcpu, or
             hw.physicalcpu_max be used instead.

```

**Comment by marius311** [comment]
Great, thanks very much for checking. So on Mac seems the correct thing is `hw.physicalcpu_max` (and def via the C API). I'll aim for that, and will likely need some help testing.

**Comment by lfield** [comment]
Work is being done on branch [phys_cores](https://github.com/BOINC/boinc/tree/phys_cores).

- [ ] Client detection
  - [x] Linux
  - [ ] Windows
  - [ ] Mac
- [x] Receive info on server, update DB, and display in host info pages
- [ ] Add planclass (?) to allow limiting to physical cores
- [ ] Use this info in scheduling

**Comment by AenBleidd** [comment]
With the release of [BUDA (Docker)](https://github.com/BOINC/boinc/wiki/BUDA-overview) that is quite stable now and already used by [BOINC Central](https://boinc.berkeley.edu/central/) and [LHC@home](https://lhcathome.cern.ch/lhcathome/index.php) we stop an active maintenance of the VBox-based applications.
Support of the VBox-based applications will not be dropped, but only the critical issues will be fixed without any new features implementation.
Thus this ticket is closed without any further actions.
